synthesizer/process/src/lib.rs

```diff
@@ -203,29 +329,68 @@ impl<N: Network> Process<N> {
         &self.universal_srs
     }
 
-    /// Returns `true` if the process contains the program with the given ID.
+    /// Returns `true` if the process or storage contains the program with the given ID.
     #[inline]
     pub fn contains_program(&self, program_id: &ProgramID<N>) -> bool {
-        self.stacks.contains_key(program_id)
+        // Check if the program is in memory.
+        if self.contains_program_in_memory(program_id) {
+            return true;
+        }
+        // Retrieve the stores.
+        if let Some(store) = self.store.as_ref() {
+            let transaction_store = store.transaction_store();
+            let deployment_store = transaction_store.deployment_store();
+            // Check if the program ID exists in the storage.
+            match deployment_store.find_transaction_id_from_program_id(program_id) {
+                Ok(Some(_)) => return true,
+                Ok(None) => debug!("Program ID not found in storage"),
+                Err(err) => debug!("Could not retrieve transaction ID for program ID: {err}"),
+            }
+        }
+
+        false
+    }
+
+    /// Returns `true` if the process contains the program with the given ID.
+    #[inline]
+    pub fn contains_program_in_memory(&self, program_id: &ProgramID<N>) -> bool {
+        // Check if the program ID is 'credits.aleo'.
+        if self.credits.as_ref().map_or(false, |stack| stack.program_id() == program_id) {
+            return true;
+        }
+        // Check if the program ID exists in the cache.
+        self.stacks.lock().contains(program_id)
     }
 
     /// Returns the stack for the given program ID.
     #[inline]
-    pub fn get_stack(&self, program_id: impl TryInto<ProgramID<N>>) -> Result<&Arc<Stack<N>>> {
+    pub fn get_stack(&self, program_id: impl TryInto<ProgramID<N>>) -> Result<Arc<Stack<N>>> {
         // Prepare the program ID.
         let program_id = program_id.try_into().map_err(|_| anyhow!("Invalid program ID"))?;
-        // Retrieve the stack.
-        let stack = self.stacks.get(&program_id).ok_or_else(|| anyhow!("Program '{program_id}' does not exist"))?;
-        // Ensure the program ID matches.
-        ensure!(stack.program_id() == &program_id, "Expected program '{}', found '{program_id}'", stack.program_id());
-        // Return the stack.
-        Ok(stack)
+        // Check if the program is 'credits.aleo'.
+        if program_id == ProgramID::<N>::from_str("credits.aleo")? {
+            return self.credits.clone().ok_or_else(|| anyhow!("Failed to get stack for 'credits.aleo'"));
+        }
+        // Try to retrieve the stack from the LRU cache.
+        if let Some(stack) = self.stacks.lock().get(&program_id) {
+            // Ensure the program ID matches.
+            ensure!(
+                stack.program_id() == &program_id,
+                "Expected program '{}', found '{program_id}'",
+                stack.program_id()
+            );
```

Comment:
is a mismatch possible? I'd expect this to be a development-only safeguard

Reply:
Agreed, removed it.